src/Icicle/Common/Exp/Simp/ANormal.hs

```diff
@@ -53,18 +57,19 @@ anormal a_fresh xx
 -- * a superset is allowed, as it will just make renaming more pessimistic
 -- * new fresh variables don't need to be mentioned, as they will not appear elsewhere as binders or mentioned
 --
-anormalAllVars :: (Hashable n, Eq n) => a -> Exp (a, Set (Name n)) n p -> Fresh n (Exp (a, Set (Name n)) n p)
+anormalAllVars :: (Hashable n, Eq n) => a -> Exp (Ann a n) n p -> Fresh n (Exp (Ann a n) n p)
 anormalAllVars a_fresh xx
- = do   (bs, x)  <- pullSubExps a_fresh xx
+ = do   (bs, x)  <- {-# SCC anormal_pullsubexps #-}pullSubExps a_fresh xx
         -- Get the union of all the variables
-        let allNames = varsOfLets bs x
+        let !allNames = {-# SCC anormal_varOfLets #-}varsOfLets bs x
         -- and rename the outside binds if they are used
-        (bs',x') <- renames allNames bs [] x
+        (bs',x') <- {-# SCC anormal_renames #-}renames allNames bs [] x
 
         -- Tag the result with the union of the original bindings (bs)
         -- as well as the new names of the bindings.
-        let allNames' = allNames <> Set.fromList (fmap fst bs')
-        return $ makeLets (a_fresh, allNames') bs' x'
+        let !allNames' = {-# SCC anormal_allnames #-} allNames <> Set.fromList (fmap fst bs')
+        let !ret = {-# SCC anormal_mklets#-}makeLets (a_fresh, allNames') bs' x'
+        return ret
 
  where
   -- All the variables inside the bindings
```

Comment:
there are a few more opportunities here: renames does a `makeLets` then `takeLets` dance, with a single substitution at each step.
you could probably give renames a substitution environment to apply to each binding, just adding to the environment and substituting the whole environment, each time. and you could build up `seen` in reverse order or something, I guess.